### PR TITLE
[WIP] Harfbuzz based Text Layout Engine integration.

### PR DESCRIPTION
**Pull request by phunsukwangdu**
This pr will focus on integrating harfbuzz based text layout engine for kivy.
For text layout engine basics plz refer [my pdf](https://github.com/phunsukwangdu/python_pango/blob/master/TextLayoutEngine.pdf)
#### There are two approach to solve the problem.
  *  use qahirah(python bindings for cairo),python_freetype,harfpy
  *  use pycairo,make python bindings for freetype(),and default python bindings provided by harfbuzz [checkout my installation guides](https://github.com/phunsukwangdu/python_pango/blob/master/Python_HarfBuzz_Setup.pdf)

#### Both of them have their own advantages and disadvantages.
* **Using Qahirah and Harfpy.**
    * Advantages:-
         1. Qahirah has access to Cairo’s support for FreeType fonts. It has
a built-in minimal FreeType wrapper.
         2. Good no of examples available online.
   * Disadvantages:-
     1. Going with this approach whole ecosystem of  libraries are required.
ie dependent on Harfpy,Qahirah,python_freetype and pybidi.
     2. Python bindings are designed for python3 only.
     3. No docs :(
* **Using pycairo and harfbuzz bindings**
  * Advantages:- 
    1. Good amount of documentations and example for pycairo.
    2. pycairo has python2 and python3 support .
  * Disadvantages:-
       1. very less doc and examples of Harfbuzz :(
       2. pycairo has no support for freetype bindings.[link](https://cairographics.org/documentation/pycairo/2/reference/text.html#class-freetypefontface-fontface)
#### Basic Skeleton of harfbuzz based text engine:-
* setup freetype font face for harfbuzz
* create harfbuzz buffer
* add text to harfbuzz buffer.
* call harfbuzz shape on buffer
* setup freetype fontface with cairo
* render glyphs with cairo
* I have worked out various harfbuzz examples  [ here](https://github.com/phunsukwangdu/python_pango/tree/master/harfbuzz_programs)
* there are lots of smaller things, such as segmentation, bidi, font selection and fallback to make harfbuzz work properly.

I am in frequent contact with harfbuzz collaborators,lurk around their irc channel,subscribed to their mail list and have browsed through their last one year mail archive.Still i know i might have made mistake somewhere here and there.
I request developers ,specially @akshayaurora :) to kindly look at my pr and guide and hint me if i am going on correct path.

###### Once this pr gets integrated Kivy will be the first in the world to get harfbuzz text layout engine (python based).

### CODE REVIEW

File: kivy/core/text/text_harfbuzz.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    def _render_end(self):
+        self.ctx.set_scaled_font(label_font)
+        self.ctx.set_scaled_font(self.lines["font"])
+        self.ctx.show_glyphs(tuple(qah.offset_glyphs(self.lines["text"], Vector(0, 0))))
+        # Imgdata = (qah.ImageSurface.get_data()) unable to implement! 
+        # only this line  is setback? we need to implement this method somehow.
+        # https://www.cairographics.org/manual/cairo-Image-Surfaces.html#cairo-image-surface-get-data
+        # https://cairographics.org/documentation/pycairo/2/reference/surfaces.html#class-imagesurface-surface
+        # img.flush().write_to_png("output.png")
```
Comment (saketkumar95) [inline kivy/core/text/text_harfbuzz.py]: @phunsukwangdu i think these all comment lines shouldn't be here in the file.

Comment (phunsukwangdu) [inline kivy/core/text/text_harfbuzz.py]: Hi @saketkumar95 thanks for picking this out.I have put this on purpose to help other developers to spot out problems and provide recommendations. Meanwhile i am too sorting this out. :)

File: kivy/core/text/text_harfbuzz.py [hunk 2]
```diff
@@ -31,10 +31,10 @@ def _get_font_id(self):
                 'italic', 'underline', 'strikethrough')])
 
     def _get_font(self):
-        return qah.get_ft_lib()
+        return qah.get_ft_lib() 
 
     def get_extents(self, text):
-        pass
+        return (100,100)     #hardcoded to get this feature working,experimantal
```
Comment (phunsukwangdu) [inline kivy/core/text/text_harfbuzz.py]: *experimental

File: kivy/core/text/text_harfbuzz.py [hunk 3]
```diff
@@ -75,18 +75,18 @@ def _render_text(self, text, x, y):
...
                 .set_source_colour(Colour.grey(0))
             )
 
     def _render_end(self):
-        self.ctx.set_scaled_font(label_font)
+        self.ctx.set_scaled_font(self.label_font)
         self.ctx.set_scaled_font(self.lines["font"])
         self.ctx.show_glyphs(tuple(qah.offset_glyphs(self.lines["text"], Vector(0, 0))))
-        Imgdata = img.data #exprimental fix
+        Imgdata = qah.ImageSurface.data #exprimental fix,still not working
```
Comment (phunsukwangdu) [inline kivy/core/text/text_harfbuzz.py]: tried `self.img.data` but of no use.Returns int but we expect buffer?

Comment (phunsukwangdu) [inline kivy/core/text/text_harfbuzz.py]: https://github.com/ldo/qahirah/blob/master/qahirah.py#L3650

**Comment by khaledhosny** [comment]
> Harfpy cannot be operated alone and to workout whole ecosystem of other libraries are required.
ie it is dependent on Qahirah,python_freetype and pybidi

Harfpy can be used without Qahirah, python_freetype or pybidi.

**Comment by phunsukwangdu** [comment]
@khaledhosny i feel honoured,u reviewed my works.For past month i was following you,kelvin13 and  Behdad Esfahbod  works and mails very closely.I really liked your quran typesetter repo.
* U get me wrong here,by Harfpy i meant ,going with the 1st approach (i will edit the description)
* second we will not be using 1st approach (harfpy one) because it is on ctypes i suppose.
* I read the mails,even kelvin13 was against using harfpy,based on similar reasons.[mail](https://lists.freedesktop.org/archives/harfbuzz/2016-August/005746.html)

@khaledhosny i request you to guide me throughout the process,till this pr gets merged.

**Comment by phunsukwangdu** [comment]
Closing,anyone who wants to come up with text shaping and text bidi can refer my [proposal](https://github.com/phunsukwangdu/Gsoc-proposal-2017/blob/master/Gsocproposal.pdf).It was my two months research and might be helpful here.

**Comment by phunsukwangdu** [comment]
Closing,anyone who wants to come up with text shaping and text bidi can refer my [proposal](https://github.com/phunsukwangdu/Gsoc-proposal-2017/blob/master/Gsocproposal.pdf).It was my two months research and might be helpful here.

**Pull request closed without merging**